### ISSUE TITLE
How to build a simple linear regression model in MyGrad?

### ISSUE THREAD

**Issue by fightthepower** [rsokl/MyGrad#213]
I am trying to make a simple linear regressin model with MyGrad. Problem is the loss value is not decreasing further. Any idea how to build a simple linear regression in MyGrad. This is my full code

```python
from sklearn.datasets import make_regression
import matplotlib.pyplot as plt
import numpy as np
import mygrad as mg

learning_rate = 0.01
X, y= make_regression(n_samples=100, n_features=1, noise=10, bias=20)

# Scatter plot of regression data
plt.scatter(X,y)
plt.show()
plt.close()

W = mg.Tensor([1,1]) # Weight, bias parameters
data_plus_bias = np.append(X,np.ones(X.shape),axis=1)

for i in range(1000):
    out = mg.matmul(data_plus_bias,W) # Dot product
    loss = ((y - out)**2).mean() # Mean square error
    loss.backward() # Gradient calculation
    W = W - learning_rate * W.grad # Weight update
    W.null_gradients()
    if i%10==0:
        print("Loss",loss.data.tolist())
```

**Comment by fightthepower** [comment]
Found the problem its because of not adding learning rate.

**Comment by rsokl** [comment]
It looks like you did a fine job - your learned linear model fits the generated data nicely!

![image](https://user-images.githubusercontent.com/29104956/75600428-47c92d80-5a7d-11ea-9447-81bb6cc860d5.png)

Keep in mind that you ought not expect your loss to go to 0: there will always be a discrepancy between your noisy data and your linear model, and the MSE will not permit these discrepancies to cancel.

To double check this, you can set the noise in your generated data to 0, and the loss of your trained model will approach 0.

A friendly note: [`noggin`](https://noggin.readthedocs.io/en/latest/) is a simple library that you can use to plot your training metrics instead of print them (assuming that you are coding in a Jupyter notebook). You need only change 3 lines of code to get real-time logging/plotting of your training metrics.

```python
from sklearn.datasets import make_regression
import numpy as np

import mygrad as mg
from noggin import create_plot

import matplotlib.pyplot as plt
%matplotlib notebook

plotter, fig, ax = create_plot(metrics="loss")

X, y = make_regression(n_samples=100, n_features=1, noise=10, bias=20)

W = mg.Tensor([1, 1])  # Weight, bias parameters
data_plus_bias = np.append(X, np.ones(X.shape), axis=1)
```
```python
learning_rate = 0.01
for i in range(1000):
    out = mg.matmul(data_plus_bias, W)  # Dot product
    loss = ((y - out) ** 2).mean()  # Mean square error
    loss.backward()  # Gradient calculation
    W = W - learning_rate * W.grad  # Weight update
    W.null_gradients()
    plotter.set_train_batch(dict(loss=loss.item()), batch_size=1)

plotter.plot()
```

**Comment by fightthepower** [comment]
Thank you for the fast reply, will optimizers like adam, sgd will be added to MyGrad?

> A friendly note: noggin is a simple library that you can use to plot your training metrics instead of print them (assuming that you are coding in a Jupyter notebook).

Do noggin work for other ide's other than jupyter notebook? I dont use jupyter notebook but an editor for python programming.

**Comment by rsokl** [comment]
[mynn](https://github.com/davidmascharka/MyNN) provides optimizers, layers, etc to use with MyGrad. We will be merging MyNN into MyGrad eventually - MyGrad 2.0 will contain all of this functionality.

You can use [noggin to log your data and to create a static plot of it](https://noggin.readthedocs.io/en/latest/how_to_log.html). This will work in any IDE. Unfortunately the live plotting only works in a Jupyter notebook.
